module/zfs/abd.c

```diff
@@ -247,9 +346,11 @@ static inline abd_t *
 abd_alloc_struct(size_t chunkcnt)
 {
 	size_t size = offsetof(abd_t, abd_u.abd_scatter.abd_chunks[chunkcnt]);
-	abd_t *abd = kmem_alloc(size, KM_PUSHPAGE);
+	abd_t *abd = kmem_zalloc(size, KM_PUSHPAGE);
 	ASSERT3P(abd, !=, NULL);
 	ABDSTAT_INCR(abdstat_struct_size, size);
+	abd->abd_create_time = gethrtime();
+	mutex_init(&abd->abd_mutex, NULL, MUTEX_DEFAULT, NULL);
```

Comment:
Sure, noted that on IRC. 

Two observations: 

Firstly, the abd structure is variable sized, and kmem_alloc always gives back same-sized allocations by design.     You can work around that with multiple caches as in zio.c at some cost, or just eat the extra memory of always giving back largest-sized abd structures.   Neither is very appealing.

Secondly, the work that lck_mtx_init() does in osfmk/i386/locks_i386.c is (semi-surprisingly) very small.   As long as we aren't doing spl mutex debugging and xnu isn't doing LCK_ATTR_DEBUG, mutex_init()'s big costs are approximately three atomic adds.

But of course if other platforms' mutex_init() is really expensive and we want this mechanism on those platforms, then we could have abds share a set of mutexes that are mutex_init()ed only once; we'd randomly (or round-robin or whatever) assign each abd one of the mutexes.  The downside is that lock contention is expensive on all platforms, and we won't want many quick abd operations waiting behind one that makes a time-consuming memory allocation where each operation is on a different abd.

(Thinking aloud: the worst case s a single lock that gives exclusive mutation access to any abd structure.   Simple, but you'll have a hotly contended lock.    So we'd want the number of locks to be ~ the number of HW threads and we'd also want to avoid false sharing, which we get for free with one lock per abd because the abd headers are bigger than cache lines.   We'd still need to keep a pointer to the appropriate mutex in each abd, so we don't even shrink the memory footprint.   Plus we contend on or wait for the spreading operator or try to be clever.)